compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/FunctionReference.kt

```diff
@@ -33,14 +33,40 @@ public sealed class FunctionReference {
 
   public abstract val fqName: FqName
   public abstract val declaringClass: ClassReference
+  public val name: String get() = fqName.shortName().asString()
 
   public val module: AnvilModuleDescriptor get() = declaringClass.module
 
   public abstract val annotations: List<AnnotationReference>
+  public abstract val parameters: List<ParameterReference>
 
   public abstract fun isAbstract(): Boolean
   public abstract fun visibility(): Visibility
 
+  /**
+   * The return type can be null for generic types like `T`. In this case try to resolve the
+   * return type with [resolveGenericReturnOrNull].
+   */
+  public abstract fun returnTypeOrNull(): ClassReference?
+
+  public fun returnType(): ClassReference = returnTypeOrNull()
+    ?: throw AnvilCompilationExceptionFunctionReference(
+      functionReference = this,
+      message = "Unable to get the return type for function $fqName."
+    )
+
+  public abstract fun resolveGenericReturnOrNull(
+    implementingClass: ClassReference.Psi
```

Comment:
I'm assuming using Psi was necessary here but wondering if there's a way to just use `ClassReference` since this is for the generic `FunctionReference`

Reply:
It's required, because the `resolveTypeReference` function requires PSI.